### PR TITLE
Windows: Fix reference counting in graphdriver

### PR DESCRIPTION
**Pull request by darstahl**
Signed-off-by: Darren Stahl <darst@microsoft.com>

**- What I did**

Fixed graphdriver reference counting for mounted layers on Windows

**- How I did it**

Windows graphdriver made the assumption that Unmount was not called on a failed Mount. This resulted in extra failures and bad reference counting when PrepareLayer failed. Found while investigating https://github.com/docker/docker/issues/27588

**- How to verify it**

In the case of a PrepareLayer failure, manual verification shows that UnprepareLayer is no longer called (and as such no longer fails)

/cc @jhowardmsft @jstarks

### CODE REVIEW

**Comment by jhowardmsft** [comment]
LGTM

**Comment by friism** [comment]
It seems like a non-trivial number of people are hitting this. @thaJeztah do you think it would be possible to prioritize this for inclusion in 17.05?

**Comment by cpuguy83** [comment]
Maybe @johnstep could take a look?

**Comment by thaJeztah** [comment]
ping @johnstep PTAL

**Comment by jhowardmsft** [comment]
@justincormack - @johnstep is on PTO this week. Any chance you could take a look?

**Comment by thaJeztah** [comment]
ach, yes you're right :smile:

**Comment by vieux** [comment]
@johnstep should be back before we freeze we should be good :)

File: daemon/graphdriver/windows/windows.go [hunk 1]
```diff
@@ -351,16 +351,13 @@ func (d *Driver) Get(id, mountLabel string) (string, error) {
 	// Getting the layer paths must be done outside of the lock.
 	layerChain, err := d.getLayerChain(rID)
 	if err != nil {
-		d.ctr.Decrement(rID)
```
Comment (johnstep) [inline daemon/graphdriver/windows/windows.go]: Why are the `Decrement` calls not needed when returning an error?

Comment (darrenstahlmsft) [inline daemon/graphdriver/windows/windows.go]: Fixed.

File: daemon/graphdriver/windows/windows.go [hunk 2]
```diff
@@ -402,9 +401,15 @@ func (d *Driver) Put(id string) error {
 		return nil
 	}
 	d.cacheMu.Lock()
+	_, exists := d.cache[rID]
 	delete(d.cache, rID)
```
Comment (darrenstahlmsft) [inline daemon/graphdriver/windows/windows.go]: Talked about it offline, both have the same effect, so leaving as is.

**Review by johnstep — APPROVED** [review]
LGTM

**Comment by vieux** [comment]
👍

**Pull request merged**